lib/route/cls/mall.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * lib/route/cls/mall.c		match-all classifier
+ *
+ *	This library is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU Lesser General Public
+ *	License as published by the Free Software Foundation version 2.1
+ *	of the License.
+ *
+ * Copyright (c) 2017 Volodymyr Bendiuga <volodymyr.bendiuga@gmail.com>
+ */
+
+/**
+ * @ingroup cls
+ * @defgroup cls_mall Match-all Classifier
+ *
+ * @{
+ */
+
+#include <netlink-private/netlink.h>
+#include <netlink-private/tc.h>
+#include <netlink/netlink.h>
+#include <netlink/attr.h>
+#include <netlink/utils.h>
+#include <netlink-private/route/tc-api.h>
+#include <netlink/route/classifier.h>
+#include <netlink/route/cls/matchall.h>
+#include <netlink/route/action.h>
+
+
+#define MALL_ATTR_CLASSID 0x01
+#define MALL_ATTR_FLAGS   0x02
+#define MALL_ATTR_ACTION  0x03
+
+
+static struct nla_policy mall_policy[TCA_MATCHALL_MAX + 1] = {
+	[TCA_MATCHALL_CLASSID]	= { .type = NLA_U32 },
+	[TCA_MATCHALL_FLAGS]	= { .type = NLA_U32 },
+};
+
+/**
+ * @name Attribute Modifications
+ * @{
+ */
+
+int rtnl_mall_set_classid(struct rtnl_cls *cls, uint32_t classid)
+{
+	struct rtnl_mall *mall;
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	mall->m_classid = classid;
+	mall->m_mask |= MALL_ATTR_CLASSID;
+
+	return 0;
+}
+
+int rtnl_mall_get_classid(struct rtnl_cls *cls, uint32_t *classid)
+{
+	struct rtnl_mall *mall;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
```

Comment:
could we use `rtnl_tc_data_peek` to avoid allocating the memory? Then, also the failure case is `-NLE_EINVAL` instead of `-NLE_NOMEM`.

Reply:
Sure, thx for advice.